src/parse_util.cpp

```diff
@@ -167,6 +170,7 @@ static int parse_util_locate_cmdsub(const wchar_t *in, const wchar_t **begin, co
         } else {
             escaped = false;
         }
+        is_token_begin = !escaped && (iswspace(*pos) || *pos == L'(');
```

Comment:
This criteria is missing some other characters that separate tokens, like `;`.
`tok_is_string_character()` should give the complete answer (the API is a bit awkward, sorry).

(Using that, we also might be able to avoid the `- 1` in `comment_end(pos) - 1`)

Reply:
Thank, but I believe `;` does not separate comment tokens. Here's an example:

```fish
echo a  # ; echo b
```

According to the existing tokenizer.cpp code that finds the end of a comment token (which I extracted to `comment_end()`), the comment token should span all the way to a `L'\n'` or `L'\0'`: https://github.com/fish-shell/fish-shell/blob/964b7a729a7ea50485b9ce56861f5e922ac26a2a/src/tokenizer.cpp#L540-L545

As for the `- 1`, it's there because at the end of the loop iteration, `pos` is incremented (`pos++`) before being checked (`*pos != L'\0'`). Without the `- 1`, `pos` is incremented beyond a null delimiter, and then `*pos` is an an out-of-bounds read.

Reply:
I meant something else: a # will also start a token in `echo foo;#bar`, hence the check is to narrow

example test case:

```fish
echo (;#"
      echo baz)
# CHECK: baz
```

Reply:
I see - thanks for catching that! I'll fix that.